### ISSUE TITLE
Deserialization of date inputs

### ISSUE THREAD

**Issue by Mr-F** [Pylons/colander#337]
I've recently found what I'm wondering is an unexpected behaviour with the deseralization code for the Date type.  Before I submitted a pull request I thought I wanted to have a short discussion, in case I was completely missing the point.

_Note: Please accept my apologies about the length of this description, I wasn't sure how much information to put in to try and cover everything_

Here is an example schema to work with:

```
class TestSchema(colander.MappingSchema):
    date_input = colander.SchemaNode(
        colander.Date()
    )
```
Running the following works as expected
```
TestSchema().deserialize({"date_input": "2020-01-01"})                                                                                                                                                                                                                                                                                               
> {'date_input': datetime.date(2020, 1, 1)}
```
and running the following raises a `colander.Invalid` exception as expected:
```
TestSchema().deserialize({"date_input": "2020/01/01"})    
```

However, if we defined a format for the date node i.e.
```
class TestSchema(colander.MappingSchema):
    date_input = colander.SchemaNode(
        colander.Date(format="%d/%m/%Y")
    )
```
Then things behave slightly differently.  A correct response is as you would expect successfully parsed and return

```
TestSchema().deserialize({"date_input": "01/01/2020"})                                                                                                                                                                                                                                                                                               
> {'date_input': datetime.date(2020, 1, 1)}
```
But incorrectly formatted strings, don't raise a `colander.Invalid` exception as I would have expected, but allows a `ValueError` to propogate. 
```
TestSchema().deserialize({"date_input": "01-01-2020"})     
...
ValueError: time data '01-01-2020' does not match format '%d/%m/%Y'               
```
Further more if the type of the input is also not a string, then the `strptime` function raises a `TypeError`
```
TestSchema().deserialize({"date_input": 12})
...
TypeError: strptime() argument 1 must be str, not int
```                                                            

In both of the cases I feel it should raise an Invalid exception in the same way that the standard ISO parsedate flow works.  If so then we just need to modify this block of code
```
try:
    if self.format:
        result = datetime.datetime.strptime(cstruct, self.format)
    else:
        result = iso8601.parse_date(cstruct)
    result = result.date()
except iso8601.ParseError as e:
    raise Invalid(
        node, _(self.err_template, mapping={'val': cstruct, 'err': e})
    )
```

### My solution        

Looking at the code I think my solution for the incorrect format would be to add an except clause to the existing try-except block.  

For the second issue with the incorrect type, there are a number of options which we could use.  Instead of adding another except block I wonder if we simply add a test e.g. not isinstance(cstruct, str), and raise an Invalid exception if required.  The other option which could be combined with the previous would be to add a flag to the construction which will indicate if we should attempt to cast the value to a string.  My thinking about this is that if we are calling `strptime` we know that it needs a string value, so having the ability to attempt the cast might make sense.  I guess you could argue that this should be handled by a preparer but that feels a bit disconnected to required that.

Anyway again please accept my apologies about the long description, but hopefully this will explain what I'm thinking.  What do people think?

If it helps I can make a pull request so you can see what I'm thinking in terms of code changes.

**Comment by stevepiercy** [comment]
Is this the block of code in Colander to which you refer?

https://github.com/Pylons/colander/blob/c91a8e6b45f9bd9afa8d3c2b87fc5369fe946b1e/src/colander/__init__.py#L1876-L1889

I think that your suggestion would be a useful addition, as it seems that this was not considered in a [recent feature addition](https://docs.pylonsproject.org/projects/colander/en/latest/changes.html#id2). A maintainer may have another opinion. Ping @bertjwregeer @mmerickel

**Comment by Mr-F** [comment]
> Is this the block of code in Colander to which you refer?
> 
> https://github.com/Pylons/colander/blob/c91a8e6b45f9bd9afa8d3c2b87fc5369fe946b1e/src/colander/__init__.py#L1876-L1889
> 
> I think that your suggestion would be a useful addition, as it seems that this was not considered in a [recent feature addition](https://docs.pylonsproject.org/projects/colander/en/latest/changes.html#id2). A maintainer may have another opinion. Ping @bertjwregeer @mmerickel

Yes, thanks for linking the code.  That is exactly the code which I think needs a small adjustment.  Just I've been working with some data recently and the error handling of deseralization wasn't behaving as I expected and traced it down to this.

**Comment by mmerickel** [comment]
I agree I think the error handling is an oversight and ValueError/TypeError should be caught in both DateTime and Date types consistently.

**Comment by stevepiercy** [comment]
Closed by #338
